### PR TITLE
Replace SymIntTable with directly storing pointer in SymInt

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82468

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: c10/core/SymInt.h [hunk 1]
```diff
@@ -31,9 +29,52 @@ using SymIntNode = c10::intrusive_ptr<SymIntNodeImpl>;
 // named data_.
 class C10_API SymInt {
  public:
+  // TODO: this needs to only accept integers, not pointers
   /*implicit*/ SymInt(int64_t d) : data_(d){};
   SymInt() = default;
 
+  // TODO: these implementations are bad
```
Comment (albanD) [inline c10/core/SymInt.h:36]: Why?

File: c10/core/SymInt.h [hunk 2]
```diff
@@ -31,9 +29,52 @@ using SymIntNode = c10::intrusive_ptr<SymIntNodeImpl>;
...
+  SymInt(const SymInt& s) : data_(0) {
+    if (s.is_symbolic()) {
+      *this = SymInt::toSymInt(s.toSymIntNodeImpl());
+    } else {
+      data_ = s.data_;
+    }
+  }
+  SymInt(SymInt&& s) : data_(s.data_) {
+    s.data_ = 0;
```
Comment (albanD) [inline c10/core/SymInt.h:45]: Is 0 a valid empty state for this?

File: c10/core/SymInt.h [hunk 3]
```diff
@@ -31,9 +29,52 @@ using SymIntNode = c10::intrusive_ptr<SymIntNodeImpl>;
...
+  SymInt& operator=(const SymInt& s) {
+    if (s.is_symbolic()) {
+      *this = SymInt::toSymInt(s.toSymIntNodeImpl());
+    } else {
+      data_ = s.data_;
+    }
+    return *this;
+  }
+  SymInt& operator=(SymInt&& s) {
+    data_ = s.data_;
```
Comment (albanD) [inline c10/core/SymInt.h:57]: This is actually moving the intrusive ptr reference ownership under the hood?

File: c10/core/SymIntNodeImpl.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#include <c10/core/SymInt.h>
+#include <c10/core/SymIntNodeImpl.h>
+
+namespace c10 {
+
+c10::SymInt SymIntNodeImpl::toSymInt() {
+  // We will need to figure out a way
+  // to dedup nodes
```
Comment (albanD) [inline c10/core/SymIntNodeImpl.cpp:8]: nodes means SymIntNodeImpl here?

File: c10/core/SymInt.h (hunk 1 above)
Comment (ezyang) [reply c10/core/SymInt.h:36]: They allocate a temporary object and then use the move constructor

File: c10/core/SymInt.h (hunk 2 above)
Comment (ezyang) [reply c10/core/SymInt.h:45]: Yes, it represents zero integer, which is the default initialization for int

File: c10/core/SymInt.h (hunk 3 above)
Comment (ezyang) [reply c10/core/SymInt.h:57]: yes, that's why s has to be cleared.

File: c10/core/SymIntNodeImpl.cpp (hunk 4 above)
Comment (ezyang) [reply c10/core/SymIntNodeImpl.cpp:8]: Actually I think this comment is obsolete now

File: c10/core/SymInt.h (hunk 1 above)
Comment (albanD) [reply c10/core/SymInt.h]: Could you update the comment to reflect this for future readers? :D

**Review by albanD — APPROVED** [review]
Ok!

**Pull request closed without merging**